sklearn/decomposition/incremental_pca.py

```diff
@@ -204,42 +208,45 @@ def partial_fit(self, X, y=None):
         else:
             self.n_components_ = self.n_components
 
-        if (self.components_ is not None) and (self.components_.shape[0]
-                                               != self.n_components_):
+        if (self.components_ is not None) and (self.components_.shape[0] !=
+                                               self.n_components_):
             raise ValueError("Number of input features has changed from %i "
                              "to %i between calls to partial_fit! Try "
-                             "setting n_components to a fixed value." % (
-                                 self.components_.shape[0], self.n_components_))
+                             "setting n_components to a fixed value." %
+                             (self.components_.shape[0], self.n_components_))
 
-        if self.components_ is None:
-            # This is the first pass through partial_fit
+        # This is the first partial_fit
+        if not hasattr(self, 'n_samples_seen_'):
             self.n_samples_seen_ = 0
-            col_var = X.var(axis=0)
-            col_mean = X.mean(axis=0)
+            self.mean_ = .0
+            self.var_ = .0
+
+        # Update stats - they are 0 if this is the fisrt step
+        col_mean, col_var, n_total_samples = \
+            _batch_mean_variance_update(X, last_mean=self.mean_,
+                                        last_variance=self.var_,
+                                        last_sample_count=self.n_samples_seen_)
+
+        # Whitening
+        if self.n_samples_seen_ == 0:
+            # If it is the first step, simply whiten X
             X -= col_mean
```

Comment:
Just a note that anytime after the first iteration, X is no longer the
original data but rather a stacked version of the existing basis and the
minibatch. However I am in agreement with this change, and am surprised
this originally passed the tests!



On Wed, Sep 9, 2015 at 7:13 AM, Olivier Grisel <notifications@github.com>
wrote:

> In sklearn/decomposition/incremental_pca.py
> <https://github.com/scikit-learn/scikit-learn/pull/5104#discussion_r39030368>
> :
>
> >              self.n_samples_seen_ = 0
> > -            col_var = X.var(axis=0)
> > -            col_mean = X.mean(axis=0)
> > +            self.mean_ = .0
> > +            self.var_ = .0
> > +
> > +        # Update stats - they are 0 if this is the fisrt step
> > +        col_mean, col_var, n_total_samples = \
> > +            _batch_mean_variance_update(X, last_mean=self.mean_,
> > +                                        last_variance=self.var_,
> > +                                        last_sample_count=self.n_samples_seen_)
> > +
> > +        # Whitening
> > +        if self.n_samples_seen_ == 0:
> > +            # If it is the first step, simply whiten X
> >              X -= col_mean
>
> I had not noticed that in @kastnerkyle <https://github.com/kastnerkyle>'s
> original PR but this is a problem: we should not silently modify the input
> data inplace.
>
> I am in favor of adding a copy=True kwarg to the IncrementalPCA
> constructor and here change the code to do:
>
> if self.copy:
>     X = X - col_meanelse:
>     X -= col_mean
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/5104/files#r39030368>.
>